### PR TITLE
8324817: Parallel GC does not pre-touch all heap pages when AlwaysPreTouch enabled and large page disabled

### PR DESCRIPTION
**Pull request by sandlerwang**
AlwaysPreTouch requires all freshly committed pages to be pre-touched. While currently PS GC does not pre-touch heap pages with -XX:+AlwaysPreTouch.
It is related to [JDK-8315923](https://bugs.openjdk.org/browse/JDK-8315923), which fixes the issue when huge pages are used. But the bug still stands if regular page are used. On linux we can reproduce this bug when /sys/kernel/mm/transparent_hugepage/enabled is madvise or never, but cannot reproduce when it is always.

A simple way to reproduce, please make sure large pages are NOT used.
Just run a test with the following parameters
-XX:+AlwaysPreTouch -Xms31g -Xmx31g -XX:+UseParallelGC
There is no pre-touching during heap initialing stage. After initialization, RSS of the test process is much smaller than memory committed.
On the contrary, using -XX:+UseG1GC and run the test again
-XX:+AlwaysPreTouch -Xms31g -Xmx31g -XX:+UseG1GC
it takes several seconds to pre-touch heap pages during initialization, and RSS usage after initialization is similar to memory committed. This is the expected behavior of AlwaysPreTouch

This issue related to [JDK-8283935](https://bugs.openjdk.org/browse/JDK-8283935), which uses alignment() instead of os::vm_page_size() as pre-touching step size. The value of alignment() is usually bigger than OS page size, which causes most heap pages are not pre-touched.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324817](https://bugs.openjdk.org/browse/JDK-8324817): Parallel GC does not pre-touch all heap pages when AlwaysPreTouch enabled and large page disabled (**Bug** - P3)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [df96a217](https://git.openjdk.org/jdk/pull/17610/files/df96a21795f8b185ac506c17cdb8a288f43dda7d)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [6a238ac4](https://git.openjdk.org/jdk/pull/17610/files/6a238ac4954998ae8e1ffcfa3846f185c168a6ed)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17610/head:pull/17610` \
`$ git checkout pull/17610`

Update a local copy of the PR: \
`$ git checkout pull/17610` \
`$ git pull https://git.openjdk.org/jdk.git pull/17610/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17610`

View PR using the GUI difftool: \
`$ git pr show -t 17610`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17610.diff">https://git.openjdk.org/jdk/pull/17610.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17610#issuecomment-1914226844)

### CODE REVIEW

**Comment by albertnetymk** [comment]
> On linux we can reproduce this bug when /sys/kernel/mm/transparent_hugepage/enabled is madvise or never, but cannot reproduce when it is always.

Could you explain why this bug is affected by that OS config flag? As you pointed out, `alignment()` (is `GenAlignment` as I read the code) can be different from `os::vm_page_size()`, so pretouching uses the wrong page-size. However, I don't get how `transparent_hugepage` is related here.

A semi-related issue, if `alignment()` can have the wrong page-size, does it mean `numa_setup_pages` (a few lines above) also needs revision?

**Comment by sandlerwang** [comment]
> > On linux we can reproduce this bug when /sys/kernel/mm/transparent_hugepage/enabled is madvise or never, but cannot reproduce when it is always.
> 
> Could you explain why this bug is affected by that OS config flag? As you pointed out, `alignment()` (is `GenAlignment` as I read the code) can be different from `os::vm_page_size()`, so pretouching uses the wrong page-size. However, I don't get how `transparent_hugepage` is related here.
> 
> A semi-related issue, if `alignment()` can have the wrong page-size, does it mean `numa_setup_pages` (a few lines above) also needs revision?

This bug is affected by that OS config flag because [JDK-8315923](https://bugs.openjdk.org/browse/JDK-8315923) added check for the config flag, and set page_size to os::vm_page_size() when the flag is always. But page_size is still alignment() when the flag is madivise or never.
Please check src/hotspot/os/linux/os_linux.cpp os::pd_pretouch_memory. In this function if HugePages::thp_mode() equals THPMode::always, page_size is set to os::vm_page_size();

For numa_setup_pages, because numa_setup_pages sets memory for from/to/eden regions, I think alignment() is sufficient here. There is some discussion about page size in numa_setup_pages here, please check. https://github.com/openjdk/jdk/pull/8090

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package gc.parallel;
+
+/**
+ * @test TestAlwaysPreTouchBehavior
+ * @summary Tests AlwaysPreTouch Bahavior, pages of java heap should be pretouched with AlwaysPreTouch enabled. This test reads RSS of test process, which should be bigger than heap size(1g) with AlwaysPreTouch enabled.
+ * @requires os.family == "linux"
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:29]: I believe ` * @requires vm.gc.Parallel` is also required here.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    if (rss == EXCEPTION_VALUE) {
+        System.out.println("cannot get RSS, just skip");
+        return; // Did not get avaiable RSS, just ignore this test
+    } else if (rss < base) {
+        System.out.println("RSS = " + rss + " smaller than committed heap memory");
+    } else {
+        System.out.println("Passed RSS = " + rss + " base value " + base);
+    }
+    Asserts.assertTrue(rss >= base, "heap rss should be bigger than committed heap mem");
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:94]: Could this be just `Asserts.assertTrue(rss >= base, "RSS: " + rss + " should be >= Committed Heap Mem: " + base);`?

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        } catch (Exception e) {
+           return EXCEPTION_VALUE;
+        }
+
+    }
+    public static void main(String [] args) {
+    long rss = 0;
+    Runtime runtime = Runtime.getRuntime();
+    long committedMemory = (runtime.totalMemory()) / 1024; // in kb
+    long base = (long)(committedMemory * 0.9);
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:79]: What's the meaning of `0.9` here? Why can one use `committedMemory` directly?

**Review by albertnetymk — COMMENTED** [review]
> Please check src/hotspot/os/linux/os_linux.cpp os::pd_pretouch_memory. In this function if HugePages::thp_mode() equals THPMode::always, page_size is set to os::vm_page_size();

Thank you for the clarification.

> For numa_setup_pages, because numa_setup_pages sets memory for from/to/eden regions, I think alignment() is sufficient here.

I dug into the `numa_setup_pages`, and it seems that `page_size` is treated as alignment, not os-page-size... (The naming here is quite misleading.)

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 2 above)
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java]: I think it is OK to use committed heap mem. Test updated

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 3 above)
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java]: 0.9 removed, just use  committedMemory directly in the updated test

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+       rss = EXCEPTION_VALUE;
+    }
+    if (rss == EXCEPTION_VALUE) {
+        System.out.println("cannot get RSS, just skip");
+        return; // Did not get avaiable RSS, just ignore this test
+    } else if (rss < committedMemory) {
+        System.out.println("RSS = " + rss + " smaller than committed heap memory " + committedMemory);
+    } else {
+        System.out.println("Passed RSS = " + rss + " committed memory " + committedMemory);
+    }
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:92]: Why these branches/print here? Could these println be incorporated in the assert msg below?

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                }
+            }
+            stringBuilder.deleteCharAt(stringBuilder.length() - 1);
+            reader.close();
+
+            String content = stringBuilder.toString();
+            return Long.valueOf(content.split("\\s+")[1].trim());
+        } catch (Exception e) {
+           return EXCEPTION_VALUE;
+        }
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:72]: This is identical to what caller does. I believe try-catch can be omitted in this method (callee), and keep the only the try-catch in the caller.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    }
+    public static void main(String [] args) {
+    long rss = 0;
+    Runtime runtime = Runtime.getRuntime();
+    long committedMemory = (runtime.totalMemory()) / 1024; // in kb
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:78]: The `()` seems not needed. Also, one doesn't need introduce a tmp `runtime` var.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package gc.parallel;
+
+/**
+ * @test TestAlwaysPreTouchBehavior
+ * @summary Tests AlwaysPreTouch Bahavior, pages of java heap should be pretouched with AlwaysPreTouch enabled. This test reads RSS of test process, which should be bigger than heap size(1g) with AlwaysPreTouch enabled.
+ * @requires vm.gc.Serial & os.family == "linux" & os.maxMemory > 2G
```
Comment (albertnetymk) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:29]: It should be Parallel, not Serial. The GC requirement should probably be on its own line to distinguish btw `vm` and `os` constraint..

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 5 above)
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java]: Thanks. Redundant try-catch removed.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 6 above)
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java]: Redundant () removed.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 7 above)
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java]: That a real problem. Serial was changed to Parallel in a separate line.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+package gc.parallel;
+
+/**
+ * @test TestAlwaysPreTouchBehavior
+ * @summary Tests AlwaysPreTouch Bahavior, pages of java heap should be pretouched with AlwaysPreTouch enabled. This test reads RSS of test process, which should be bigger than heap size(1g) with AlwaysPreTouch enabled.
+ * @requires vm.gc.Parallel
+ * @requires os.family == "linux" & os.maxMemory > 2G
+ * @library /test/lib
```
Comment (tschatzl) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:31]: The test only works in release builds as the debug builds incidentally always pretouch bases.
I.e. add `@requires vm.debug != true`.
There does not seem to be a `@requires vm.release` or something.

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 8 above)
Comment (tschatzl) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:31]: Another option is to `@require vm.debug` and add `-XX:-ZapUnusedHeapArea` to the run options ( suggested by @albertnetymk ).

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    long committedMemory = runtime.totalMemory() / 1024; // in kb
+    try {
+       rss = getProcessRssInKb();
+    } catch (Exception e) {
+       rss = EXCEPTION_VALUE;
+    }
+    if (rss == EXCEPTION_VALUE) {
+        System.out.println("cannot get RSS, just skip");
+        return; // Did not get avaiable RSS, just ignore this test
+    }
```
[the remark this answers is not in the document]
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:83]: Thanks, EXCEPTION_VALUE  related code were deleted. Test updated

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+        String content = stringBuilder.toString();
+        return Long.valueOf(content.split("\\s+")[1].trim());
+    }
+    public static void main(String [] args) {
+    long rss = 0;
+    Runtime runtime = Runtime.getRuntime();
+    long committedMemory = runtime.totalMemory() / 1024; // in kb
+    try {
+       rss = getProcessRssInKb();
```
[the remark this answers is not in the document]
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:76]: Inconsistent issue fixed

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java [hunk 11]
```diff
@@ -48,36 +49,27 @@
 import java.io.*;
 
 public class TestAlwaysPreTouchBehavior {
-    static final long EXCEPTION_VALUE = -1L;
     public static long getProcessRssInKb() throws IOException {
         String pid = ManagementFactory.getRuntimeMXBean().getName().split("@")[0];
         // Read RSS from /proc/$pid/status. Only avaiable on Linux
```
Comment (tschatzl) [inline test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java:54]:
```suggestion
        // Read RSS from /proc/$pid/status. Only available on Linux.
```

**Review by tschatzl — APPROVED** [review]
Please fix the typo I did not find last time before integrating.

Lgtm.

File: test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java (hunk 11 above)
Comment (sandlerwang) [reply test/hotspot/jtreg/gc/parallel/TestAlwaysPreTouchBehavior.java]: Of course. Typo fixed, thanks

**Pull request closed without merging**